src/Compilers/Core/Portable/Operations/BasicBlock.cs

```diff
@@ -1,152 +1,119 @@
-            _predecessors = ImmutableHashSet.CreateBuilder<BasicBlock>();
-        }
-
-        public BasicBlockKind Kind { get; private set; }
-        public ImmutableArray<IOperation> Statements => _statements.ToImmutable();
-
-        /// <summary>
-        /// PROTOTYPE(dataflow): Tuple is temporary return type, we probably should use special structure instead.
-        /// </summary>
-        public (IOperation Condition, bool JumpIfTrue, Branch Branch) Conditional => InternalConditional;
-
         /// <summary>
-        /// PROTOTYPE(dataflow): During CR there was a suggestion to use different name - "Successor".
+        /// PROTOTYPE(dataflow): Rename this type to "BasicBlockBuilder" and source file to "ControlFlowGraphBuilder.BasicBlockBuilder.cs"
         /// </summary>
-        public (IOperation Value, Branch Branch) Next => InternalNext;
-
-        public ImmutableHashSet<BasicBlock> Predecessors => _predecessors.ToImmutable();
+        internal sealed class BasicBlock
+        {
+            private readonly ArrayBuilder<IOperation> _statementsBuilder;
+            private readonly HashSet<BasicBlock> _predecessorsBuilder;
+            public (IOperation Condition, bool JumpIfTrue, Branch Branch) InternalConditional;
+            public (IOperation Value, Branch Branch) InternalNext;
+            
+            public BasicBlock(BasicBlockKind kind)
+            {
+                Kind = kind;
+                _statementsBuilder = ArrayBuilder<IOperation>.GetInstance();
+                _predecessorsBuilder = new HashSet<BasicBlock>();
+            }
 
-        public int Ordinal { get; internal set; } = -1;
+            public IReadOnlyCollection<IOperation> Statements => _statementsBuilder;
+            public ImmutableArray<IOperation> GetImmutableStatements() => _statementsBuilder.ToImmutableAndFree();
+            public ISet<BasicBlock> Predecessors => _predecessorsBuilder;
+            public BasicBlockKind Kind { get; }
 
-        public bool IsReachable { get; internal set; } = false;
+            public int Ordinal { get; set; } = -1;
 
-        /// <summary>
-        /// Enclosing region
-        /// </summary>
-        public ControlFlowGraph.Region Region { get; internal set; }
+            public bool IsReachable { get; set; } = false;
 
-        internal void AddStatement(IOperation statement)
-        {
-            _statements.Add(statement);
-        }
+            public ControlFlowRegion Region { get; set; }
 
-        internal void AddStatements(ImmutableArray<IOperation> statements)
-        {
-            _statements.AddRange(statements);
-        }
+            internal void AddStatement(IOperation statement)
+            {
+                _statementsBuilder.Add(statement);
+            }
 
-        internal void RemoveStatements()
-        {
-            _statements.Clear();
-        }
+            internal void AddStatements(IEnumerable<IOperation> statements)
```

Comment:
Consider moving this and `Statements` to use ArrayBuilder<T> instead, as this method will end up using Linq under the hood and allocating multiple iterators.

Reply:
> Consider moving this and Statements to use ArrayBuilder<T> instead, as this method will end up using Linq under the hood and allocating multiple iterators.

I'll follow  up on that myself

---
In reply to: [192249332](https://github.com/dotnet/roslyn/pull/27254#discussion_r192249332) [](ancestors = 192249332)